File: app/src/main/java/ui/adapter/TodoAdapter.java
package ui.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.AppCompatCheckBox;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.AsyncListDiffer;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import hlv.cute.todo.databinding.ItemTodoBinding;
import model.Todo;
import utils.TextHelper;

public class TodoAdapter extends RecyclerView.Adapter<TodoAdapter.ViewHolder> {

    private final Context context;
    private final AsyncListDiffer<Todo> differ;

    private final OnCheckChangedListener onCheckChangedListener;
    private final OnClickMenuListener onClickMenuListener;

    public TodoAdapter(Context context, OnCheckChangedListener onCheckChangedListener, OnClickMenuListener onClickMenuListener) {
        this.context = context;
        this.onCheckChangedListener = onCheckChangedListener;
        this.onClickMenuListener = onClickMenuListener;

        DiffUtil.ItemCallback<Todo> diffCallback = new DiffUtil.ItemCallback<Todo>() {
            @Override
            public boolean areItemsTheSame(@NonNull Todo oldItem, @NonNull Todo newItem) {
                return oldItem.getId() == newItem.getId();
            }

            @Override
            public boolean areContentsTheSame(@NonNull Todo oldItem, @NonNull Todo newItem) {
                return oldItem.compareTo(newItem) == 0;
            }
        };

        differ = new AsyncListDiffer<>(this, diffCallback);
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemTodoBinding binding = ItemTodoBinding.inflate(LayoutInflater.from(context), parent, false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.bind(differ.getCurrentList().get(position), onCheckChangedListener, onClickMenuListener);
    }

    @Override
    public int getItemCount() {
        return differ.getCurrentList().size();
    }

    static class ViewHolder extends RecyclerView.ViewHolder {

        private final AppCompatCheckBox checkBox;
        private final AppCompatImageView imgMenu;
        private final TextView txtCategory;

        private final ConstraintLayout lytDate;
        private final TextView txtDate, txtClock;

        private final TextView txtLow, txtNormal, txtHigh;

        public ViewHolder(@NonNull ItemTodoBinding binding) {
            super(binding.getRoot());

            checkBox = binding.aChkBoxTitle;
            imgMenu = binding.aImgMenu;
            txtCategory = binding.txtCategory;

            lytDate = binding.lytDate;
            txtDate = binding.txtDate;
            txtClock = binding.txtClock;

            txtLow = binding.txtLowPriority;
            txtNormal = binding.txtNormalPriority;
            txtHigh = binding.txtHighPriority;
        }

        private void bind(Todo todo, OnCheckChangedListener onCheckChangedListener, OnClickMenuListener onClickMenuListener) {
            //must be set null
            //to avoid when recyclerview scroll, default implemented interface body called!
            checkBox.setOnCheckedChangeListener(null);

            //for first run
            checkBox.setChecked(todo.isDone());
            checkBox.setText(todo.getTitle());

            if (todo.isDone())
                TextHelper.addLineThrough(checkBox);
            else
                TextHelper.removeLineThrough(checkBox);

            checkBox.setOnCheckedChangeListener((compoundButton, checked) -> {
                //after checked
                if (compoundButton.isPressed()) {
                    todo.setDone(checked);

                    if (todo.isDone())
                        TextHelper.addLineThrough(checkBox);
                    else
                        TextHelper.removeLineThrough(checkBox);

                    onCheckChangedListener.onChange(todo.getId());
                }
            });


            imgMenu.setOnClickListener(view -> onClickMenuListener.onClick(todo));

            if (todo.getCategory() != null) {
                txtCategory.setVisibility(View.VISIBLE);
                txtCategory.setText(todo.getCategory());
            } else {
                txtCategory.setVisibility(View.GONE);
            }

            if (todo.getArriveDate() != 0) {
                lytDate.setVisibility(View.VISIBLE);
                txtDate.setText(todo.getDateTime().getPersianDate());
                txtClock.setText(todo.getClock());
            } else {
                lytDate.setVisibility(View.GONE);
            }

            switch (todo.getPriority()) {
                case LOW:
                default:
                    txtLow.setVisibility(View.VISIBLE);
                    txtNormal.setVisibility(View.GONE);
                    txtHigh.setVisibility(View.GONE);
                    break;
                case NORMAL:
                    txtNormal.setVisibility(View.VISIBLE);
                    txtLow.setVisibility(View.GONE);
                    txtHigh.setVisibility(View.GONE);
                    break;
                case HIGH:
                    txtHigh.setVisibility(View.VISIBLE);
                    txtLow.setVisibility(View.GONE);
                    txtNormal.setVisibility(View.GONE);
                    break;
            }
        }
    }

    public AsyncListDiffer<Todo> getDiffer() {
        return differ;
    }

    public interface OnCheckChangedListener {
        void onChange(int todoID);
    }

    public interface OnClickMenuListener {
        void onClick(Todo todo);
    }
}

File: app/src/main/java/ui/adapter/TodoAdapter.kt
package ui.adapter

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Point
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.CompoundButton
import androidx.recyclerview.widget.AsyncListDiffer
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import hlv.cute.todo.databinding.ItemTodoBinding
import model.Priority
import model.Todo
import utils.TextHelper

class TodoAdapter(
    private val context: Context,
    private val onCheckChangedListener: (todoID: Int, oldValueDone: Boolean?) -> Unit,
    private val onClickMenuListener: (todo: Todo, anchor: View, wholeItem: View, coordinatePoint: Point?) -> Unit,
    private val onClickMoreListener: (todoMore: Todo) -> Unit,
) : RecyclerView.Adapter<TodoAdapter.ViewHolder>() {

    val differ: AsyncListDiffer<Todo>

    init {
        val diffCallback: DiffUtil.ItemCallback<Todo> = object : DiffUtil.ItemCallback<Todo>() {
            override fun areItemsTheSame(oldItem: Todo, newItem: Todo): Boolean {
                return oldItem.id == newItem.id
            }

            override fun areContentsTheSame(oldItem: Todo, newItem: Todo): Boolean {
                return oldItem.compareTo(newItem) == 0
            }
        }

        differ = AsyncListDiffer(this, diffCallback)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = ItemTodoBinding.inflate(
            LayoutInflater.from(
                context
            ), parent, false
        )
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(
            differ.currentList[position],
            onCheckChangedListener,
            onClickMenuListener,
            onClickMoreListener
        )
    }

    override fun getItemCount(): Int {
        return differ.currentList.size
    }

    inner class ViewHolder(private val binding: ItemTodoBinding) :
        RecyclerView.ViewHolder(binding.root) {

        @SuppressLint("ClickableViewAccessibility")
        fun bind(
            todo: Todo,
            onCheckChangedListener: (todoID: Int, oldValueDone: Boolean?) -> Unit,
            onClickMenuListener: (todo: Todo, anchor: View, wholeItem: View, coordinatePoint: Point?) -> Unit,
            onClickMoreListener: (todoMore: Todo) -> Unit
        ) {
            //must be set null
            //to avoid when recyclerview scroll, default implemented interface body called!
            binding.aChkBoxTitle.setOnCheckedChangeListener(null)

            //root.setTransitionName("t-" + todo.getId());

            //for first run
            binding.aChkBoxTitle.isChecked = todo.isDone

            binding.aChkBoxTitle.text = todo.title

            if (todo.isDone)
                TextHelper.addLineThrough(binding.aChkBoxTitle) else TextHelper.removeLineThrough(
                binding.aChkBoxTitle
            )

            binding.aChkBoxTitle.setOnCheckedChangeListener { compoundButton: CompoundButton, checked: Boolean ->
                //after checked
                if (compoundButton.isPressed) {
                    val oldValueDone = todo.isDone

                    todo.isDone = checked


                    if (todo.isDone)
                        TextHelper.addLineThrough(binding.aChkBoxTitle)
                    else
                        TextHelper.removeLineThrough(binding.aChkBoxTitle)

                    onCheckChangedListener(todo.id, oldValueDone)
                }
            }

            var rawClickPoint: Point? = null
            binding.root.setOnTouchListener { _, event ->
                if (event.actionMasked == MotionEvent.ACTION_DOWN)
                    rawClickPoint = Point(event.rawX.toInt(), event.rawY.toInt())

                return@setOnTouchListener false
            }

            binding.root.setOnClickListener {
                onClickMenuListener(
                    todo,
                    it,
                    binding.root,
                    rawClickPoint
                )
            }

            binding.aImgMenu.setOnClickListener {
                onClickMenuListener(
                    todo,
                    it,
                    binding.root,
                    null
                )
            }

            if (todo.category != null) {
                binding.txtCategory.visibility = View.VISIBLE
                binding.txtCategory.text = todo.category
            } else {
                binding.txtCategory.visibility = View.GONE
            }

            if (todo.arriveDate != 0L) {
                binding.lytDate.visibility = View.VISIBLE
                binding.txtDate.text = todo.dateTime!!.persianDate
                binding.txtClock.text = todo.clock
            } else {
                binding.lytDate.visibility = View.GONE
            }

            when (todo.priority) {
                Priority.LOW -> {
                    binding.txtLowPriority.visibility = View.VISIBLE
                    binding.txtNormalPriority.visibility = View.GONE
                    binding.txtHighPriority.visibility = View.GONE
                }

                Priority.NORMAL -> {
                    binding.txtNormalPriority.visibility = View.VISIBLE
                    binding.txtLowPriority.visibility = View.GONE
                    binding.txtHighPriority.visibility = View.GONE
                }

                Priority.HIGH -> {
                    binding.txtHighPriority.visibility = View.VISIBLE
                    binding.txtLowPriority.visibility = View.GONE
                    binding.txtNormalPriority.visibility = View.GONE
                }

                else -> {
                    binding.txtLowPriority.visibility = View.VISIBLE
                    binding.txtNormalPriority.visibility = View.GONE
                    binding.txtHighPriority.visibility = View.GONE
                }
            }

            binding.root.postOnAnimation {
                if (binding.aChkBoxTitle.lineCount >= 3) {
                    binding.txtMore.visibility = View.VISIBLE
                } else {
                    binding.txtMore.visibility = View.GONE
                }
            }

            binding.txtMore.setOnClickListener {
                onClickMoreListener(todo)
            }
        }

    }
}